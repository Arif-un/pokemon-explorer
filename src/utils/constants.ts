import bug from '@/assets/type-icons/bug.svg'
import dark from '@/assets/type-icons/dark.svg'
import dragon from '@/assets/type-icons/dragon.svg'
import electric from '@/assets/type-icons/electric.svg'
import fairy from '@/assets/type-icons/fairy.svg'
import fighting from '@/assets/type-icons/fighting.svg'
import fire from '@/assets/type-icons/fire.svg'
import flying from '@/assets/type-icons/flying.svg'
import ghost from '@/assets/type-icons/ghost.svg'
import grass from '@/assets/type-icons/grass.svg'
import ground from '@/assets/type-icons/ground.svg'
import ice from '@/assets/type-icons/ice.svg'
import normal from '@/assets/type-icons/normal.svg'
import poison from '@/assets/type-icons/poison.svg'
import psychic from '@/assets/type-icons/psychic.svg'
import rock from '@/assets/type-icons/rock.svg'
import steel from '@/assets/type-icons/steel.svg'
import water from '@/assets/type-icons/water.svg'

export const DEFAULT_POKEMON_FETCH_LIMIT = 20
export const DEFAULT_POKEMON_FETCH_OFFSET = 0

export const TYPE_ICONS = {
  bug,
  dark,
  dragon,
  electric,
  fairy,
  fighting,
  fire,
  flying,
  ghost,
  grass,
  ground,
  ice,
  normal,
  poison,
  psychic,
  rock,
  steel,
  water
}

export const TYPE_COLORS = {
  normal: '#A8A77A',
  fire: '#EE8130',
  water: '#6390F0',
  electric: '#F7D02C',
  grass: '#7AC74C',
  ice: '#96D9D6',
  fighting: '#C22E28',
  poison: '#A33EA1',
  ground: '#E2BF65',
  flying: '#A98FF3',
  psychic: '#F95587',
  bug: '#A6B91A',
  rock: '#B6A136',
  ghost: '#735797',
  dragon: '#6F35FC',
  dark: '#705746',
  steel: '#B7B7CE',
  fairy: '#D685AD'
}

export type PokemonType = keyof typeof TYPE_COLORS

export const POKEMON_STATS = {
  hp: { max: 255, color: '#ff0000' },
  attack: { max: 190, color: '#ff0097' },
  defense: { max: 250, color: '#11e800' },
  'special-attack': { max: 194, color: '#ff4d00' },
  'special-defense': { max: 250, color: '#383bff' },
  speed: { max: 200, color: '#00a9ff' }
}
