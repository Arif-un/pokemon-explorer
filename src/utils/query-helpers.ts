import { type QueryKey } from '@tanstack/react-query'

import pokemonPlaceholder from '@/assets/Pokeball.svg'
import { type Pokemon, type PokemonDetailsResponse, type PokemonResponse } from '@/types/Pokemon'

import { POKEMON_STATS, TYPE_COLORS, TYPE_ICONS } from './constants'

export interface Stats {
  name: string
  value: number
  max: number
  color: string
}

export interface PokemonTypes {
  name?: string
  icon?: string
  color?: string
}

export function pickPokemonImage(pokemonDetails: Pokemon | PokemonDetailsResponse) {
  const images = Array.isArray(pokemonDetails?.sprites)
    ? pokemonDetails.sprites[0]?.sprites
    : pokemonDetails?.sprites

  return (
    images?.other?.['official-artwork']?.front_shiny ||
    images?.other?.['official-artwork']?.front_default ||
    images?.other?.dream_world?.front_default ||
    images?.other?.home?.front_shiny ||
    images?.front_shiny ||
    pokemonPlaceholder
  )
}

export function preparePokemonDetails(pokemonDetails: PokemonDetailsResponse) {
  const id = pokemonDetails?.id
  const image = pickPokemonImage(pokemonDetails)
  const types = pokemonDetails?.types?.map<PokemonTypes>(typeInfo => ({
    name: typeInfo?.type?.name,
    icon: typeInfo?.type?.name && TYPE_ICONS[typeInfo.type.name],
    color: typeInfo?.type?.name && TYPE_COLORS[typeInfo.type.name]
  }))

  const abilities = pokemonDetails?.abilities?.map(abilityInfo => abilityInfo?.ability?.name)

  const baseStats = pokemonDetails?.stats?.reduce<Stats[]>((acc, statInfo) => {
    if (statInfo?.stat?.name) {
      const statName = statInfo.stat.name
      acc.push({
        name: statName,
        value: statInfo.base_stat || 0,
        max: POKEMON_STATS[statName].max,
        color: POKEMON_STATS[statName].color
      })
    }
    return acc
  }, [])

  return { id, image, types, abilities, baseStats }
}

export function transformTypeFromCacheData(data: { pokemon_v2_type: { name: string } }[]) {
  return data.map(item => ({
    type: {
      name: item.pokemon_v2_type.name
    }
  }))
}

function findPokemonInCache(
  queriesData: [QueryKey, PokemonResponse | undefined][],
  pokemonName: string
) {
  for (const queryData of queriesData) {
    const cachedPokemon = queryData?.[1]?.pokemons?.find(pokemon => pokemon.name === pokemonName)
    if (cachedPokemon) return cachedPokemon
  }
  return null
}

export function preparePokemonDetailsInitialData(
  name: string,
  cachedPokemonQueries: [QueryKey, PokemonResponse | undefined][]
) {
  const cachedPokemon = findPokemonInCache(cachedPokemonQueries, name)

  if (!cachedPokemon) return

  const { sprites, types } = cachedPokemon
  const transformedSprites = sprites?.[0]?.sprites || null
  const transformedTypes = transformTypeFromCacheData(types)

  return {
    sprites: transformedSprites,
    types: transformedTypes
  } as PokemonDetailsResponse
}
