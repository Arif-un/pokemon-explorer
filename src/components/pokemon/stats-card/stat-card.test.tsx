import { render, screen } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'

import { type Stats } from '@/utils/query-helpers'

import StatsCard from './stats-card'

vi.mock('framer-motion', async () => {
  const actual = await vi.importActual('framer-motion')
  return {
    ...actual,
    AnimatePresence: ({
      children,
      onExitComplete
    }: {
      children: React.ReactNode
      onExitComplete?: () => void
    }) => {
      onExitComplete?.()
      return <div>{children}</div>
    }
  }
})

vi.mock('@/animations/common-animation', () => ({
  fadeAnimation: {}
}))

vi.mock('@/animations/stat-card-animation', () => ({
  startCardParentAnimations: {},
  statsPieItemAnimations: {}
}))

describe('StatsCard Component', () => {
  const baseStats: Stats[] = [
    { name: 'HP', value: 60, max: 100, color: 'red' },
    { name: 'Attack', value: 80, max: 100, color: 'blue' },
    { name: 'Defense', value: 70, max: 100, color: 'green' }
  ]

  it('renders correctly with baseStats', () => {
    const finishNavigation = vi.fn()
    const { asFragment } = render(
      <StatsCard baseStats={baseStats} isNavigating={false} finishNavigation={finishNavigation} />
    )

    // Check if the component renders the stats
    baseStats.forEach(stat => {
      expect(screen.getByText(stat.name)).toBeInTheDocument()
      expect(screen.getByText(stat.value.toString())).toBeInTheDocument()
    })

    // Check if StatPie component renders correctly without mocking
    const statPies = screen.getAllByRole('img')
    expect(statPies).toHaveLength(baseStats.length)

    baseStats.forEach((stat, index) => {
      const percent = (stat.value / stat.max) * 100
      expect(statPies[index].getAttribute('aria-label')).toContain(`${stat.name}: ${percent}%`)
    })

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders nothing while navigating', () => {
    const finishNavigation = vi.fn()
    const { asFragment } = render(
      <StatsCard baseStats={baseStats} isNavigating={true} finishNavigation={finishNavigation} />
    )

    // Expect the Stats header not to appear
    expect(screen.queryByText('Stats')).not.toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })

  it('calls finishNavigation on animation exit complete', () => {
    const finishNavigation = vi.fn()
    render(<StatsCard baseStats={baseStats} isNavigating={false} finishNavigation={finishNavigation} />)

    // Ensure that finishNavigation was called
    expect(finishNavigation).toHaveBeenCalledTimes(1)
  })
})
