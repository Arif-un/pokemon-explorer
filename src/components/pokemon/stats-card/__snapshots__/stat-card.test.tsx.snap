// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatsCard Component > renders correctly with baseStats 1`] = `
<DocumentFragment>
  <section>
    <div>
      <h2
        class="text-slate-600 dark:text-slate-500 mb-1"
      >
        Stats
      </h2>
      <ul
        aria-label="List of Pokemon Base Stats"
        class="grid grid-cols-3 gap-y-6 gap-x-4 border dark:border-slate-800 bg-slate-100 bg-opacity-70 dark:bg-slate-900 p-4 rounded-xl"
      >
        <li>
          <div
            class="inline-flex justify-center flex-col items-center min-w-24"
          >
            <figure
              class="size-14 rounded-full bg-slate-400 bg-opacity-60 dark:bg-slate-900 grid place-content-center relative"
            >
              <div
                aria-label="HP: 60%"
                class="size-14 rounded-full"
                role="img"
              />
              <div
                class="size-11 rounded-full bg-slate-100 dark:bg-slate-900 absolute inset-0 m-auto grid place-content-center"
                style="color: red;"
              />
              <span
                aria-label="Pokemon stat name"
                aria-live="polite"
                class="absolute m-auto right-0 left-0 size-4 -bottom-2 text-xs text-slate-600 dark:text-slate-400"
              >
                60
              </span>
            </figure>
            <figcaption
              class="text-sm text-center text-slate-600 dark:text-slate-400 capitalize mt-3 font-semibold"
            >
              HP
            </figcaption>
          </div>
        </li>
        <li>
          <div
            class="inline-flex justify-center flex-col items-center min-w-24"
          >
            <figure
              class="size-14 rounded-full bg-slate-400 bg-opacity-60 dark:bg-slate-900 grid place-content-center relative"
            >
              <div
                aria-label="Attack: 80%"
                class="size-14 rounded-full"
                role="img"
              />
              <div
                class="size-11 rounded-full bg-slate-100 dark:bg-slate-900 absolute inset-0 m-auto grid place-content-center"
                style="color: blue;"
              />
              <span
                aria-label="Pokemon stat name"
                aria-live="polite"
                class="absolute m-auto right-0 left-0 size-4 -bottom-2 text-xs text-slate-600 dark:text-slate-400"
              >
                80
              </span>
            </figure>
            <figcaption
              class="text-sm text-center text-slate-600 dark:text-slate-400 capitalize mt-3 font-semibold"
            >
              Attack
            </figcaption>
          </div>
        </li>
        <li>
          <div
            class="inline-flex justify-center flex-col items-center min-w-24"
          >
            <figure
              class="size-14 rounded-full bg-slate-400 bg-opacity-60 dark:bg-slate-900 grid place-content-center relative"
            >
              <div
                aria-label="Defense: 70%"
                class="size-14 rounded-full"
                role="img"
              />
              <div
                class="size-11 rounded-full bg-slate-100 dark:bg-slate-900 absolute inset-0 m-auto grid place-content-center"
                style="color: green;"
              />
              <span
                aria-label="Pokemon stat name"
                aria-live="polite"
                class="absolute m-auto right-0 left-0 size-4 -bottom-2 text-xs text-slate-600 dark:text-slate-400"
              >
                70
              </span>
            </figure>
            <figcaption
              class="text-sm text-center text-slate-600 dark:text-slate-400 capitalize mt-3 font-semibold"
            >
              Defense
            </figcaption>
          </div>
        </li>
      </ul>
    </div>
  </section>
</DocumentFragment>
`;

exports[`StatsCard Component > renders nothing while navigating 1`] = `
<DocumentFragment>
  <section>
    <div />
  </section>
</DocumentFragment>
`;
